dbms/src/Functions/FunctionsRegexp.h

```diff
@@ -0,0 +1,1009 @@
+        }
+        else
+        {
+            // Construct actual param
+            param = new ParamStringNotNullableAndConst(col_size, col_const->getDataAt(0));
+            param_type = ParamType::StringNotNullableAndConst;
+        }
+    }
+
+    void handleStringNonConstCol(size_t col_size)
+    {
+        if (col_ptr->isColumnNullable())
+        {
+            auto nested_ptr = static_cast<const ColumnNullable &>(*(col_ptr)).getNestedColumnPtr();
+            const auto * tmp = checkAndGetColumn<ColumnString>(&(*nested_ptr));
+            const auto * null_map = &(static_cast<const ColumnNullable &>(*(col_ptr)).getNullMapData());
+
+            // Construct actual param
+            param = new ParamStringNullableAndNotConst(col_size, null_map, static_cast<const void *>(&(tmp->getChars())), static_cast<const void *>(&(tmp->getOffsets())));
+            param_type = ParamType::StringNullableAndNotConst;
+        }
+        else
+        {
+            // This is a pure string vector column
+            const auto * tmp = checkAndGetColumn<ColumnString>(&(*(col_ptr)));
+
+            // Construct actual param
+            param = new ParamStringNotNullableAndNotConst(col_size, static_cast<const void *>(&(tmp->getChars())), static_cast<const void *>(&(tmp->getOffsets())));
+            param_type = ParamType::StringNotNullableAndNotConst;
+        }
+    }
+
+    void setParamStringTypeAndGenerateParam(size_t col_size)
+    {
+        const auto * col_const = typeid_cast<const ColumnConst *>(&(*(col_ptr)));
+        if (col_const != nullptr)
+            handleStringConstCol(col_size, col_const);
+        else
+            handleStringNonConstCol(col_size);
+    }
+
+    // TODO implement it in next pr
+    void setParamIntTypeAndGenerateParam(size_t col_size [[maybe_unused]])
+    {
+        throw Exception("Not implemented so far");
+    }
+
+    ParamType param_type;
+    ColumnPtr col_ptr;
+    StringRef default_str;
+    Int64 default_int [[maybe_unused]];
+    void * param;
+};
+
+// Unifying these names is necessary in macros
+#define EXPR_PV_VAR_NAME expr_pv
+#define PAT_PV_VAR_NAME pat_pv
+#define MATCH_TYPE_PV_VAR_NAME match_type_pv
+
+#define EXPR_PARAM_PTR_VAR_NAME expr_param
+#define PAT_PARAM_PTR_VAR_NAME pat_param
+#define MATCH_TYPE_PARAM_PTR_VAR_NAME match_type_param
+
+#define RES_ARG_VAR_NAME res_arg
+
+// Unify the name of functions that actually execute regexp
+#define REGEXP_CLASS_MEM_FUNC_IMPL_NAME executeRegexpFunc
+
+// Do not merge GET_ACTUAL_STRING_PARAM and GET_ACTUAL_INT_PARAM together,
+// as this will generate more useless codes and templates.
+
+#define ENUMERATE_PARAM_VARIANT_CASES(NAME, pv_name, param_name, next_process) \
+    case ParamVariant::ParamType::NAME:                                        \
+    {                                                                          \
+        ParamVariant::Param##NAME *(param_name) = (pv_name).getParam##NAME();  \
```

Comment:
How about using a micro for this conversion? If so, `getParamXXX` function family can be removed.

Reply:
> How about using a micro for this conversion? If so, `getParamXXX` function family can be removed.

Converted pointer is contained in ParamVariant, it will add complexity when we want to convert it as we have to get it from ParamVariant. Moreover, debugging with macro is hard, I think we should avoid macro as much as possible. At last, these simple functions must be inlined by compiler and there is no performance difference between macro and function.

Reply:
How about
```suggestion
        #define ParamName(NAME) ParamVariant::Param##NAME
        #define ToParamPtr(NAME, ptr) (reinterpret_cast<ParamName(NAME) *>(ptr))
        auto *(param_name) = ToParamPtr(NAME, (pv_name).param);  \
```

Reply:
> How about this?

okk, I will resolve it.